### PR TITLE
Add primary UDN net-attach-def

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:
This PR adds support to user-defined-network binding by deploying the appropriate NetworkAttachmentDefinition. 
This is part of the process described on [kubevirt user guide](https://kubevirt.io/user-guide/network/network_binding_plugins/#definition-flow).
The behavior added is currently dev-preview. Hence, the NetworkAttachmentDefinition is not deployed by the proper component, but is deployed by ipam-extentions. This should change in future PRs.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/network/kubevirt_ipam_controller.go
Comment (oshoval) [inline pkg/network/kubevirt_ipam_controller.go:52]: if the CRD is NA, there will be a fatal https://github.com/kubevirt/cluster-network-addons-operator/pull/1841/files#diff-463a5613afbb95588d747500e304c2d348c7fce8ec35117067322e7aac36a8dcR106
so basically this isn't needed (just need to fatal if CRD is NA + conf.ipam == true)
the only advantage of it, is that we can disable it easily if needed
wdyt ?

File: hack/components/bump-kubevirt-ipam-controller.sh
Comment (RamLavi) [reply hack/components/bump-kubevirt-ipam-controller.sh]: good idea. thanks!

File: pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go
Comment (oshoval) [inline pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:525]: we should fail only if ipam is enable and CRD doesnt exists, not in all the cases
which what happens here iiuc

**Comment by RamLavi** [comment]
@oshoval do you understand why it [fails](https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/10246887390/job/28344944532?pr=1841)?

**Comment by oshoval** [comment]
> @oshoval do you understand why it [fails](https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/10246887390/job/28344944532?pr=1841)?

`error: timed out waiting for the condition on kubevirts/kubevirt`
for some reason lately it cant install kubevirt all the time

File: pkg/network/kubevirt_ipam_controller.go
Comment (RamLavi) [reply pkg/network/kubevirt_ipam_controller.go:52]: I think it was discussed that we want to fail early.. totally forgot. will fix. thanks!

**Comment by RamLavi** [comment]
> > @oshoval do you understand why it [fails](https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/10246887390/job/28344944532?pr=1841)?
> 
> `error: timed out waiting for the condition on kubevirts/kubevirt` for some reason lately it cant install kubevirt all the time
> 
> maybe it is because the wait was too fast, and wait doesnt wait for the resource to be created and then need a loop of wait with shorter timeout, or to use the new flag of kubectl wait that waits for the resource to exists [kubernetes/kubernetes#122994](https://github.com/kubernetes/kubernetes/pull/122994)
> 
> unless it is something deeper, and to know we would need logs

Looks like kubevirt/kubevirt now [wait](https://github.com/kubevirt/kubevirt/blob/649245379da5c128cd74a05c236c600326d1f827/hack/ci/entrypoint.sh#L90) 15m

File: test/check/check.go
Comment (RamLavi) [reply test/check/check.go:661]: > but why we need all this logic if we just create the nad 

This code is for CNAO CI to make sure the NAD is deployed. It's now CNAO's responsibility to deploy this NAD, and so CI should check it's deployed, and also removed. 

>and if it doesnt exists we have the error that nad doesnt exists without any additional changes?

small correction - only if the NAD's CRD doesn't exists then CNAO will fail to deploy with the error you mention. 

> is it a must because it was added here ? https://github.com/kubevirt/cluster-network-addons-operator/pull/1841/files#diff-dd3d0d54317e754fd95de914de01a73b5578626b42e755fdad4998186d552600R90

technically yeah, but it's intentional. As I mentioned above - it's CNAO's job to make sure the resources it is in charge of deploying are deployed.

File: pkg/network/kubevirt_ipam_controller_test.go
Comment (RamLavi) [reply pkg/network/kubevirt_ipam_controller_test.go:20]: it shouldn't be. fixed.

File: pkg/network/kubevirt_ipam_controller_test.go
Comment (RamLavi) [reply pkg/network/kubevirt_ipam_controller_test.go:24]: DONE

File: test/check/check.go
Comment (RamLavi) [reply test/check/check.go:351]: ah copy err. thanks!

File: pkg/network/kubevirt_ipam_controller_test.go
Comment (oshoval) [inline pkg/network/kubevirt_ipam_controller_test.go:20]: basically since all are false, those are the default values anyhow
so you can just create an empty struct
unless one of them is better to be explicit false, but this is not the case here

**Comment by oshoval** [comment]
> > > > @oshoval do you understand why it [fails](https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/10246887390/job/28344944532?pr=1841)?
> > > 
> > > 
> > > `error: timed out waiting for the condition on kubevirts/kubevirt` for some reason lately it cant install kubevirt all the time
> > > maybe it is because the wait was too fast, and wait doesnt wait for the resource to be created and then need a loop of wait with shorter timeout, or to use the new flag of kubectl wait that waits for the resource to exists [kubernetes/kubernetes#122994](https://github.com/kubernetes/kubernetes/pull/122994)
> > > unless it is something deeper, and to know we would need logs
> > 
> > 
> > Looks like kubevirt/kubevirt now [wait](https://github.com/kubevirt/kubevirt/blob/649245379da5c128cd74a05c236c600326d1f827/hack/ci/entrypoint.sh#L90) 15m
> 
> this change is 4y ago it seems, but worth to try to extend

I think the best we cant do is to install kubevirt using kind.sh
for that we need to add an option to bypass kubevirt-ipam-controller installation / ipam claim crd
or to undeploy them after installation is done
one of the reasons is that kind.sh does this when installing kubevirt (among few other little things)
```
        $OCI_BIN exec -t $node bash -c "echo 'fs.inotify.max_user_watches=1048576' >> /etc/sysctl.conf"
        $OCI_BIN exec -t $node bash -c "echo 'fs.inotify.max_user_instances=512' >> /etc/sysctl.conf"
        $OCI_BIN exec -i $node bash -c "sysctl -p /etc/sysctl.conf"
        if [[ "${node}" =~ worker ]]; then
            kubectl label nodes $node node-role.kubernetes.io/worker="" --overwrite=true
        fi
```

File: pkg/components/components.go
Comment (oshoval) [inline pkg/components/components.go:746]: do we need update ? (maybe the reconcile use it when it recreate?)

please consider moving it under the else of `allowMultus`
otherwise there are duplicate / overlapping rbacs due to multusClusterRoles in U/S, which i tried to refrain from

you can move the flag into multusClusterRoles and it will determine if to add * or to add just the diff needed please

File: pkg/network/kubevirt_ipam_controller_test.go
Comment (RamLavi) [reply pkg/network/kubevirt_ipam_controller_test.go:20]: DONE

**Comment by RamLavi** [comment]
> > > > > @oshoval do you understand why it [fails](https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/10246887390/job/28344944532?pr=1841)?
> > > > 
> > > > 
> > > > `error: timed out waiting for the condition on kubevirts/kubevirt` for some reason lately it cant install kubevirt all the time
> > > > maybe it is because the wait was too fast, and wait doesnt wait for the resource to be created and then need a loop of wait with shorter timeout, or to use the new flag of kubectl wait that waits for the resource to exists [kubernetes/kubernetes#122994](https://github.com/kubernetes/kubernetes/pull/122994)
> > > > unless it is something deeper, and to know we would need logs
> > > 
> > > 
> > > Looks like kubevirt/kubevirt now [wait](https://github.com/kubevirt/kubevirt/blob/649245379da5c128cd74a05c236c600326d1f827/hack/ci/entrypoint.sh#L90) 15m
> > 
> > 
> > this change is 4y ago it seems, but worth to try to extend
> 
> I think the best we cant do is to install kubevirt using kind.sh for that we need to add an option to bypass kubevirt-ipam-controller installation / ipam claim crd / cert-manager or to undeploy them after installation is done one of the reasons is that kind.sh does this when installing kubevirt (among few other little things) or we can also just add those in our repo
> 
> ```
>         $OCI_BIN exec -t $node bash -c "echo 'fs.inotify.max_user_watches=1048576' >> /etc/sysctl.conf"
>         $OCI_BIN exec -t $node bash -c "echo 'fs.inotify.max_user_instances=512' >> /etc/sysctl.conf"
>         $OCI_BIN exec -i $node bash -c "sysctl -p /etc/sysctl.conf"
>         if [[ "${node}" =~ worker ]]; then
>             kubectl label nodes $node node-role.kubernetes.io/worker="" --overwrite=true
>         fi
> ```

Maybe, but it should be done in a separate PR to be sure.
IIRC this is the first time that lane began to be flaky, am I correct?

**Comment by oshoval** [comment]
> Maybe, but it should be done in a separate PR to be sure. IIRC this is the first time that lane began to be flaky, am I correct?

started lately
saw it one time or so before this PR

File: pkg/components/components.go
Comment (RamLavi) [reply pkg/components/components.go:746]: > do we need update ? (maybe the reconcile use it when it recreate?)

We [do](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/pkg/apply/apply.go#L62).

> 
> please consider moving it under the else of `allowMultus` otherwise there are duplicate / overlapping rbacs due to multusClusterRoles in U/S, which i tried to refrain from
> 
> you can move the flag into multusClusterRoles and it will determine if to add * or to add just the diff needed please

OK

File: pkg/components/components.go
Comment (oshoval) [inline pkg/components/components.go:965]: those arent just for UDN
so either keep them at the original place please
or rename the function for something more generic
or just keep as you did before this change (where all the 6 are outside) please

Comment (RamLavi) [reply pkg/components/components.go:965]: suggestions for a generic name?

Comment (oshoval) [reply pkg/components/components.go:965]: sorry please see the EDIT, the option for generic name is wrong
we must keep those 3 outside please

**Comment by RamLavi** [comment]
/hold
I want to make sure there is no problem deploying this on D/S

File: pkg/components/components.go
Comment (RamLavi) [reply pkg/components/components.go:965]: OK

**Comment by oshoval** [comment]
imo i wont go into details about it being dev-preview and such (PR desc and commit),
but just say it is not deployed as part of  kubevirt ipam controller

beside that lgtm
thanks

**Comment by RamLavi** [comment]
> imo i wont go into details about it being dev-preview and such (PR desc and commit), but just say it is not deployed as part of kubevirt ipam controller
> 

why not?

> beside that lgtm thanks

thanks!

**Comment by oshoval** [comment]
> > imo i wont go into details about it being dev-preview and such (PR desc and commit), but just say it is not deployed as part of kubevirt ipam controller
> 
> why not?

We don't know when the change will happen, and imho the PR description should be clear and succinct
details like those are fine on Jira tickets, less on PR / commits

**Comment by RamLavi** [comment]
>/hold
I want to make sure there is no problem deploying this on KCLI

manually pushed this PR to my quay and deployed on KCLI:
```
# kubectl get pod -A
NAMESPACE                NAME                                                  READY   STATUS              RESTARTS   AGE
cert-manager             cert-manager-67c98b89c8-48n2r                         1/1     Running             0          5m46s
cert-manager             cert-manager-cainjector-5c5695d979-fktvh              1/1     Running             0          5m46s
cert-manager             cert-manager-webhook-7f9f8648b9-vr5b8                 1/1     Running             0          5m46s
cluster-network-addons   bridge-marker-8m82x                                   1/1     Running             0          2m
cluster-network-addons   bridge-marker-ksdkj                                   1/1     Running             0          2m
cluster-network-addons   bridge-marker-s6zw4                                   1/1     Running             0          2m
cluster-network-addons   cluster-network-addons-operator-597956cbcb-xhjd2      2/2     Running             0          8m23s
cluster-network-addons   dynamic-networks-controller-ds-5ctnx                  0/1     ContainerCreating   0          2m
cluster-network-addons   dynamic-networks-controller-ds-kc6vt                  0/1     ContainerCreating   0          2m
cluster-network-addons   dynamic-networks-controller-ds-ptjmc                  0/1     ContainerCreating   0          2m
cluster-network-addons   kube-cni-linux-bridge-plugin-c9b5j                    1/1     Running             0          2m
cluster-network-addons   kube-cni-linux-bridge-plugin-lm4bf                    1/1     Running             0          2m
cluster-network-addons   kube-cni-linux-bridge-plugin-p5l2t                    1/1     Running             0          2m
cluster-network-addons   kubemacpool-cert-manager-79579b8fd5-dz4q5             1/1     Running             0          2m
cluster-network-addons   kubemacpool-mac-controller-manager-64769cf7db-2w8zb   2/2     Running             0          2m
cluster-network-addons   kubevirt-ipam-controller-manager-7bb84b4fc8-7k2tb     1/1     Running             0          2m
cluster-network-addons   multus-6s66t                                          1/1     Running             0          2m
cluster-network-addons   multus-f6nw5                                          1/1     Running             0          2m
cluster-network-addons   multus-f84fh                                          1/1     Running             0          2m
cluster-network-addons   passt-binding-cni-dl2tg                               1/1     Running             0          2m
cluster-network-addons   passt-binding-cni-kr4j9                               1/1     Running             0          2m
cluster-network-addons   passt-binding-cni-ssswr                               1/1     Running             0          2m
cluster-network-addons   secondary-dns-8657555bd9-sngtr                        2/2     Running             0          2m
kube-system              coredns-76f75df574-j5mfz                              1/1     Running             0          26d
kube-system              coredns-76f75df574-swvgl                              1/1     Running             0          26d
kube-system              etcd-ovn-control-plane                                1/1     Running             0          26d
kube-system              kube-apiserver-ovn-control-plane                      1/1     Running             0          26d
kube-system              kube-controller-manager-ovn-control-plane             1/1     Running             0          26d
kube-system              kube-multus-ds-kcnzp                                  1/1     Running             0          26d
kube-system              kube-multus-ds-q78m8                                  1/1     Running             0          26d
kube-system              kube-multus-ds-vmx7w                                  1/1     Running             0          26d
kube-system              kube-scheduler-ovn-control-plane                      1/1     Running             0          26d
kubevirt                 virt-api-98cf864cc-q5qv2                              1/1     Running             0          26d
kubevirt                 virt-api-98cf864cc-wbmp4                              1/1     Running             0          26d
kubevirt                 virt-controller-5d65ddf8c5-4vllz                      1/1     Running             0          26d
kubevirt                 virt-controller-5d65ddf8c5-bmmml                      1/1     Running             0          26d
kubevirt                 virt-handler-d8zbf                                    1/1     Running             0          26d
kubevirt                 virt-handler-rrtjp                                    1/1     Running             0          26d
kubevirt                 virt-handler-v9v5q                                    1/1     Running             0          26d
kubevirt                 virt-operator-865f487cf6-2m8dv                        1/1     Running             0          26d
kubevirt                 virt-operator-865f487cf6-bnjcs                        1/1     Running             0          26d
local-path-storage       local-path-provisioner-7577fdbbfb-zgk6v               1/1     Running             0          26d
nmstate                  nmstate-cert-manager-88549958d-4qctp                  1/1     Running             0          26d
nmstate                  nmstate-handler-d27wz                                 1/1     Running             0          26d
nmstate                  nmstate-handler-l42pl                                 1/1     Running             0          26d
nmstate                  nmstate-handler-sl9k8                                 1/1     Running             0          26d
nmstate                  nmstate-operator-7b66695676-5j8xx                     1/1     Running             0          26d
nmstate                  nmstate-webhook-57774bfd49-d65ht                      1/1     Running             0          26d
ovn-kubernetes           ovnkube-control-plane-bc6d94b84-m2wwm                 1/1     Running             0          26d
ovn-kubernetes           ovnkube-identity-65cbc8cbff-r49nx                     1/1     Running             0          26d
ovn-kubernetes           ovnkube-node-42bvv                                    6/6     Running             0          26d
ovn-kubernetes           ovnkube-node-kfxhz                                    6/6     Running             0          26d
ovn-kubernetes           ovnkube-node-rngmq                                    6/6     Running             0          26d
ovn-kubernetes           ovs-node-626pm                                        1/1     Running             0          26d
ovn-kubernetes           ovs-node-ggs9d                                        1/1     Running             0          26d
ovn-kubernetes           ovs-node-ps6hx                                        1/1     Running             0          26d
```

CNAO successfuly deployed the NAD:
```
# kubectl get net-attach-def
NAME                           AGE
primary-user-defined-network   4m24s
```

/hold cancel

**Comment by oshoval** [comment]
the fact the ipam lane doesnt work at all, is bit risky because if this PR broke anything we wont know

**Comment by RamLavi** [comment]
> the fact the ipam lane doesnt work at all, is bit risky because if this PR broke anything we wont know worst case we can run e2e locally at least if there will be rush but maybe i found the problem

ACK. 
I think we can't move on until we are sure. opened an issue, please update there what you found. thanks!

**Comment by oshoval** [comment]
imo please try to address meanwhile sonarcloud (or to ignore them on sonar cloud itself / merge manually at the end)
and to run the e2e locally (using the whole script that git actions runs)
it seems fixing the ipam lane will take time, and i dont think we should block this PR due to that

**Comment by RamLavi** [comment]
> imo please try to address meanwhile sonarcloud (or to ignore them on sonar cloud itself / merge manually at the end) and to run the e2e locally (using the whole script that git actions runs) it seems fixing the ipam lane will take time, and i dont think we should block this PR due to that

then how do we know if this PR didn't break IPAM? 
I mean, it does touch IPAM code. 
Can you run it locally and tell for more certainty that this PR is not hurting IPAM?

**Comment by RamLavi** [comment]
> > imo please try to address meanwhile sonarcloud (or to ignore them on sonar cloud itself / merge manually at the end) and to run the e2e locally (using the whole script that git actions runs) it seems fixing the ipam lane will take time, and i dont think we should block this PR due to that
> 
> then how do we know if this PR didn't break IPAM? I mean, it does touch IPAM code. Can you run it locally and tell for more certainty that this PR is not hurting IPAM?

[update]
Of course I'm working on the lifecycle lane and sonar cloud in parallel

**Comment by oshoval** [comment]
before this PR, one time there was a flake about kubevirt install, and then with a rerun it worked
so before this PR i consider it to work,
we just need to run it locally with this PR please 

i am working on migrating the lane to prow, it will take bit time, it is your call if the above check is fine or not
(imho it is fine unless waiting wont miss the FF date, but we have time until that)

**Comment by RamLavi** [comment]
> before this PR, one time there was a flake about kubevirt install, and then with a rerun it worked so before this PR i consider it to work, we just need to run it locally with this PR please
> 
> i am working on migrating the lane to prow, it will take bit time, it is your call if the above check is fine or not (imho it is fine unless waiting wont miss the FF date, but we have time until that)

It also failed on this [PR](https://github.com/kubevirt/cluster-network-addons-operator/pull/1842) we issued yesterday, so I consider this lane failing/flaky. 

I ran it manually on my local and it passed, though I am not comfortable overriding this lane. As you are more familiar with it, I trust you if you want to override it on this PR. 
As for moving to prow instead of gitActions, I'm good with anything that will work :)

**Comment by oshoval** [comment]
this is your decision if to override it or not, i just tried to help and gave my opinion

**Comment by oshoval** [comment]
you can try to add `// NOSONAR` on the annoying file (if it can be added to one line and affect a struct instead it is better of course)

**Comment by oshoval** [comment]
you can temporary cherry-pick https://github.com/kubevirt/cluster-network-addons-operator/pull/1843 and hold if you want
it should fix the problem

**Comment by RamLavi** [comment]
> you can try to add `// NOSONAR` on the annoying file (if it can be added to one line and affect a struct instead it is better of course)
> 
> sometimes there is also a way to mark an error as accepted / false positive on the UI (at least for some types of warnings)

fixed in separate commit

File: pkg/components/components.go
Comment (oshoval) [inline pkg/components/components.go:360]: at least some of the changes here are cosmetics which i personally against (it is clearer as it was imo)
and also it is not related to the PR
please remove them, we can create a follow-up PR and discuss them there

if any logical change is needed, lets add just that (maybe it is in a commit already, but i looked on the total diff)

Comment (RamLavi) [reply pkg/components/components.go:360]: sure I'll move to a separate PR

**Comment by RamLavi** [comment]
> you can temporary cherry-pick #1843 and hold until it is merged if you want it should fix the problem

great! once it is merged I'll retest

**Comment by oshoval** [comment]
> > you can temporary cherry-pick #1843 and hold until it is merged if you want it should fix the problem
> 
> great! once it is merged I'll retest

merged, rebase please and then retest, git actions rebase isnt as good as tide is

**Comment by RamLavi** [comment]
#1843 is merged.
/retest

**Comment by oshoval** [comment]
please rebase, the current run doesnt have the updated code
i just checked to make sure

**Comment by RamLavi** [comment]
[Rebase](https://github.com/kubevirt/cluster-network-addons-operator/compare/b459b570aac6cf3d3d32068d7f086a1a668726c9..e56a15c566ecd0d5857a30d6590f4300ee520b23)

**Comment by oshoval** [comment]
how come sonar is happy now ?
anyhow sonar happy we are happy

**Comment by RamLavi** [comment]
> how come sonar is happy now ? (rebase did it maybe) anyhow sonar happy we are happy

For some reason the duplication only upped by 2.9%, not 3%.

File: test/e2e/workflow/deployment_test.go
Comment (oshoval) [inline test/e2e/workflow/deployment_test.go:37]: what about just always allow 10s ?

Comment (RamLavi) [reply test/e2e/workflow/deployment_test.go:37]: yeah I tend to agree..

Comment (oshoval) [reply test/e2e/workflow/deployment_test.go]: nit: maybe the commit message of that need to be changed please

**Comment by oshoval** [comment]
about the workflow failure - this is not a solution but maybe a temporary workaround and then we can fix it on a follow-up (imo 10s is not enough as we spoke)

please check if you want this function `deploy_cnao_cr`, and make sure multus is in the CR before the ipam one
it will save some reconcile cycles

File: test/check/check.go
Comment (oshoval) [inline test/check/check.go:400]: The condition can be moved into the function
but we can consider doing it on a follow-up for all those methods not in this PR

**Comment by RamLavi** [comment]
> about the workflow failure - this is not a solution but maybe a temporary workaround and then we can fix it on a follow-up (imo 10s is not enough as we spoke): please check if you want this function `deploy_cnao_cr`, and make sure multus is in the CR before the ipam one it will save some reconcile cycles
> 
> EDIT - if CNAO waits for processing / ready state too short, it might fail because in the middle the state becomes fail for various errors, only eventually it will be processing / ready - not sure if relevant here, but i did saw it before in areas that remind this area - to conclude we should be here more tolerant imho and eventually consistent, not too strict

The issue was not relevant to timing, something much more dumb.
Removed the timeouts, let's see how we handle without them.

File: test/check/check.go
Comment (RamLavi) [reply test/check/check.go:400]: >but we can consider doing it on a follow-up for all those methods not in this PR
+!

**Comment by oshoval** [comment]
Please let me know once all is passing and ready for review,
I think it will be better for us both.
of course if there are any questions just mention me please
Thanks

File: test/check/check.go
Comment (oshoval) [inline test/check/check.go:666]: please use
if ...; err != nil {
}

File: test/check/check.go
Comment (oshoval) [inline test/check/check.go:673]: here you can just return err basically, but one can consider it less readable
so either this please or short writing 
```
if ...; err != nil {
}
```

**Review by oshoval — COMMENTED** [review]
nits, beside that lgtm
thanks

File: test/check/check.go
Comment (RamLavi) [reply test/check/check.go:673]: [same](https://github.com/kubevirt/cluster-network-addons-operator/pull/1841/commits/f634d35bfb3cee4337321935d578c78af7f5d6cf#r1708736686)

File: test/check/check.go
Comment (oshoval) [reply test/check/check.go:666]: well this is a new function, but i wont insist

**Review by oshoval — APPROVED** [review]

File: pkg/components/components.go
Comment (oshoval) [inline pkg/components/components.go:745]: one thing we can consider
limit this to specific NS (i.e default atm)
or use it as Role instead RoleBinding with limited NS which is the default NS

**Comment by oshoval** [comment]
/lgtm cancel

please at least consider https://github.com/kubevirt/cluster-network-addons-operator/pull/1841#discussion_r1708769127
if needed we can done it on a follow-up, but if it is possible to do it in this one it is better

File: data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml
Comment (qinqon) [inline data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml:10]: Let's differenciate from nad name, since those are different 
```yaml
name: [CNAO namespace]-kubevirt-passt-binding-plugin-net
```

File: hack/components/bump-kubevirt-ipam-controller.sh
Comment (qinqon) [inline hack/components/bump-kubevirt-ipam-controller.sh:151]: ditto: Let's not reference UDN at the nad, just at the kubevirt CR config on HCO.

**Review by qinqon — CHANGES_REQUESTED** [review]
first pass let's not reference UDN here.

File: pkg/components/components.go
Comment (RamLavi) [reply pkg/components/components.go:745]: My thoughts exactly.. we can decide to use more specifc RBAC when we decide what the final resting place of this NAD

File: data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml
Comment (RamLavi) [reply data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml:5]: > [CNAO namespace]
why CNAO namespace?

File: hack/components/bump-kubevirt-ipam-controller.sh
Comment (RamLavi) [reply hack/components/bump-kubevirt-ipam-controller.sh:151]: hmm, I'm not against this, but how will anyone ever understand the connection between the NAD and the binding? shouldn't they have at least something in common? a label maybe?

File: data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml
Comment (qinqon) [reply data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml:5]: After some back and forth let's name this 
```yaml
name: primary-user-defined-network-kubevirt-binding
```

File: data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml
Comment (qinqon) [reply data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml:10]: After some back and forth let's name this 
```yaml
name: primary-user-defined-network-kubevirt-binding
```

File: hack/components/bump-kubevirt-ipam-controller.sh
Comment (qinqon) [reply hack/components/bump-kubevirt-ipam-controller.sh:151]: Let's name this 
```
004-primary-udn-kubevirt-binding-networkattachdef.yaml
```

Comment (oshoval) [reply hack/components/bump-kubevirt-ipam-controller.sh:151]: maybe just ?
004-primary-udn-kubevirt-binding-nad.yaml

File: test/check/check.go
Comment (RamLavi) [reply test/check/check.go:666]: I will make the change in a separate PR after this one is merged

Comment (oshoval) [reply test/check/check.go:666]: sure thx

**Comment by RamLavi** [comment]
[Change](https://github.com/kubevirt/cluster-network-addons-operator/compare/f634d35bfb3cee4337321935d578c78af7f5d6cf..d59c8526b0c2e8ab184e8acead8950398fbf7469): Fixes for @qinqon 's review

File: data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml
Comment (RamLavi) [reply data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml]: DONE

File: data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml
Comment (RamLavi) [reply data/kubevirt-ipam-controller/004-primary-udn-networkattachdef.yaml]: DONE

**Review by qinqon — APPROVED** [review]
/lgtm

**Comment by oshoval** [comment]
Please create a patch release, so it will be consumed wherever needed

**Pull request merged**
